File: python/array/zigzag.py
class Zigzag:
    def convert(self, s, numRows):
        if numRows < 2 or len(s) == 0:
            return s

        zs = ""

        for i in range(0, numRows):
            span1 = 2 * (numRows - i - 1)
            span2 = 2 * i

            add = True

            b = i

            while b < len(s):
                zs += s[b]

                if span1 == 0:
                    b += span2
                elif span2 == 0:
                    b += span1
                elif add:
                    b += span1
                else:
                    b += span2

                add = not add


        return zs

File: swift/array/Zigzag.swift
//
//  Zigzag.swift
//  Algorithm
//
//  Created by Haibo Yan on 3/12/18.
//  Copyright © 2018 Haibo Yan. All rights reserved.
//

import Foundation

public class Zigzag {
    func convert(_ s: String, _ numRows: Int) -> String {
        if (numRows < 2 || s.count == 0) {
            return s
        }
        var zs = ""
        let cs = s.map{String($0)}
        
        for i in 0...numRows-1 {
            let span1 = 2 * (numRows - i - 1)
            let span2 = 2 * i
            
            var add = true

            var b = i
            
            while b < s.count {
                zs += cs[b]
            
                switch (span1, span2, add) {
                case (0, _, _):
                    b += span2
                case (_, 0, _),
                     (_, _, true):
                    b += span1
                default:
                    b += span2
                }
                
                add = !add
            }
        }
        
        return zs
    }
}
